File: Programmers/Lv. 3/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    '''
    지형을 나타내는 직사각형 담긴 2차원 배열 rectangle [좌측하단x, 좌측하단y, 우측상단x, 우측상단y]
    초기 캐릭터 위치 characterX, characterY
    아이템 위치 itemX, itemY
    '''  
    moves = [(1,0), (-1,0), (0,1), (0,-1)]
    
    # 만약 ㄷ 모양 테두리가 있으면 테두리로 가지 않고 
    # 바로 올라가는 경우가 생기기 때문에 2배를 해서 field를 생성하는 것
    inField = []
    for lx, ly, rx, ry in rectangle:
        for i in range(lx*2+1, rx*2):
            for j in range(ly*2+1, ry*2):
                inField.append([i, j])
    
    outField = []
    for lx, ly, rx, ry in rectangle:
        for i in range(lx*2, rx*2+1):
            if [i, ly * 2] not in inField:
                outField.append([i, ly * 2])
            if [i, ry * 2] not in inField:
                outField.append([i, ry * 2])
        for i in range(ly*2+1, ry*2):
            if [lx * 2, i] not in inField:
                outField.append([lx*2, i])
            if [rx * 2, i] not in inField:
                outField.append([rx*2, i])
                    
    q = deque()
    q.append([[characterX*2, characterY*2], 0])
    
    visited = []
    visited.append([characterX*2, characterY*2])
    while q:
        pos, cnt = q.popleft()
        if pos == [itemX*2, itemY*2]:
            return cnt / 2
        for move in moves:
            dx = pos[0] + move[0]
            dy = pos[1] + move[1]
            if [dx, dy] not in visited and [dx, dy] in outField:
                visited.append([dx, dy])
                q.append([[dx, dy], cnt+1])

File: Programmers/Lv. 3/아이템 줍기.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> rectangle, int characterX, int characterY, int itemX, int itemY) {
    int answer = 0;
    
    int x_move[4] = {0, 0, 1, -1};
    int y_move[4] = {1, -1, 0, 0};
    
    int max_x = 0, max_y = 0;
    
    vector<vector<int>> box;    
    vector<vector<bool>> visited;    
    
    for (vector<int> rec : rectangle)
    {
        if (rec[2] > max_x)
        {
            max_x = rec[2];
        }
        if (rec[3] > max_y)
        {
            max_y = rec[3];
        }
    }
    
    for (int x = 0; x < 2*max_x + 1; x++)
    {
        vector<int> tmp;
        vector<bool> tmp2;
        for (int y = 0; y < 2*max_y + 1; y++)
        {
            tmp.push_back(0);
            tmp2.push_back(false);
        }
        box.push_back(tmp);
        visited.push_back(tmp2);
    }
    
    for (vector<int> rec : rectangle)
    {
        int x1 = 2*rec[0];
        int y1 = 2*rec[1];
        int x2 = 2*rec[2];
        int y2 = 2*rec[3];
        for (int x = x1 + 1; x < x2; x++)
        {
            for (int y = y1 + 1; y < y2; y++)
            {
                box[x][y] = -1;                
            }
        }
    }
    
    for (vector<int> rec : rectangle)
    {
        int x1 = 2*rec[0];
        int y1 = 2*rec[1];
        int x2 = 2*rec[2];
        int y2 = 2*rec[3];
        for (int x = x1; x < x2 + 1; x++)
        {
            for (int y = y1; y < y2 + 1; y++)
            {
                if (box[x][y] != -1)
                {
                    box[x][y] = 1;                    
                }                
            }
        }
    }       
    
    queue<vector<int>> q;
    q.push({2*characterX, 2*characterY, 0});
    while (!q.empty())
    {
        vector<int> tmp = q.front();
        q.pop();
        int cx = tmp[0];
        int cy = tmp[1];
        int cnt = tmp[2];
        
        if (cx == 2*itemX && cy == 2*itemY)
        {
            return int(cnt/2);
        }
        
        for (int i = 0; i < 4; i++)
        {
            int nx = cx + x_move[i];
            int ny = cy + y_move[i];
            
            if (nx >= 0 && nx <= 2*max_x && ny >= 0 && ny <= 2*max_y && box[nx][ny] == 1 && !visited[nx][ny])
            {
                visited[nx][ny] = true;
                q.push({nx, ny, cnt+1});
            }
        }
    }
    
    return answer;
}